### ISSUE TITLE
Mosaic augmentation

### ISSUE THREAD

**Issue by pfeatherstone** [ultralytics/yolov5#991]
Does the mosaic augmentation choose the cropped regions such that they contain at least one object or does it simply pick a random crop?
This is an issue in datasets where objects are sparse withing an image.

**Comment by glenn-jocher** [comment]
@pfeatherstone a random crop is chosen without consideration of object locations. Different datasets may benefit to different degrees from mosaic, so you may want to try training with and without it on your custom dataset. We've recently added mosaic to the training hyps next to mixup, so now it's easy to control the probability of any given image being mosaiced:
https://github.com/ultralytics/yolov5/blob/5a9c5c1d3b02c8796e6af153d57af5821d8db0d6/data/hyp.scratch.yaml#L32-L33

**Comment by abhiagwl4262** [comment]
@glenn-jocher, As I am looking into code, it seems.. Mixup is with mosaics only. Shouldn't it be for the actual images as well ?

**Comment by glenn-jocher** [comment]
@abhiagwl4262 yes, your observation is correct. Ideally these would be independent as you noted. Would you like to you take the lead on this and submit a PR? The relevant code in the dataloader is here:
https://github.com/ultralytics/yolov5/blob/7220cee1d1dc1f14003dbf8d633bbb76c547000c/utils/datasets.py#L519-L531

TODO: separate mixup and mosaic completely

**Comment by abhiagwl4262** [comment]
@glenn-jocher. Yeah Sure, I Will be doing experiments on weekend and get back to you with the observations.

**Comment by abhiagwl4262** [comment]
` if self.augment:
            # Augment imagespace
            if not self.mosaic:
                img, labels = random_perspective(img, labels,
                                                 degrees=hyp['degrees'],
                                                 translate=hyp['translate'],
                                                 scale=hyp['scale'],
                                                 shear=hyp['shear'],
                                                 perspective=hyp['perspective'])
`

**Comment by abhiagwl4262** [comment]
@glenn-jocher Something similar, we can see here as well.

https://github.com/ultralytics/yolov5/blob/7220cee1d1dc1f14003dbf8d633bbb76c547000c/utils/datasets.py#L554

"random_perspective" is being only in the case mosaic augmentation is not selected.

**Comment by glenn-jocher** [comment]
@abhiagwl4262 this part is ok actually, as the mosaic function includes it's own call to random_perspective. The reason it has it's own call is that it needs to pass a border in to the mosaic which the regular image does not use. The border is typically -320, and removes 320 pixels around the 1280x1280 mosaic to leave a 640x640 square.

EDIT: an alternative implemenation would simply set a zero border value for regular images, and then perhaps we could simply have one single call to random_perspective. Also note though that mixup calls random_perspective a second time for the mixed up image, so the base and the mixed image go through different augmentations, which is important.

**Comment by abhiagwl4262** [comment]
@glenn-jocher I see that mosaic is always true for training. So the flow of pre-processing during training is 

**img ->mosaic -> mosaiced_img -> augment_hsv -> flipud  -> fliplr -> bgr2rgb**

So I changed code to support use of both original and mosaic data. And I got imporovement. 



yours yolov5l  - map@0.5 - **0.8923**, map@0.5:0.95 - **0.6704**
Mine yolov5l   - map@0.5 - **0.8957** map@0.5:0.95 - **0.6742**
Mine yolo5l new best -   map@0.5 -**0.8962**    map@0.5:0.95 - **0.6765**

Ran it for 50 epoches

**Comment by abhiagwl4262** [comment]
@glenn-jocher 
Have a look at the precision graph. There is sudden dip in midway. What can be the reason for the same ? See the logs as well. Transition from 7th epoch to 8th epoch, even the loss values decreased. 

![checkpoint_voc_l_latest_repo_result](https://user-images.githubusercontent.com/22645080/94999864-00d68400-05da-11eb-94d4-cafec01c7440.png)

      0/49     22.8G   0.05801   0.01138   0.02627   0.09566       103       640         0         0 0.0006552 0.0001638   0.04856  0.009223   0.02582
      1/49     26.6G    0.0473   0.01133   0.02528   0.08392        85       640         0         0    0.0339   0.01556   0.03631  0.009078   0.02402
      2/49     26.6G   0.03846   0.01132   0.02321   0.07299       115       640    0.2968   0.02271    0.1371   0.08756   0.02784  0.007603   0.02072
      3/49     26.6G   0.03321  0.009987    0.0206   0.06379        86       640    0.4275    0.0565    0.2547    0.1427   0.02411  0.006403   0.01768
      4/49     26.6G   0.03003   0.00864   0.01835   0.05702        81       640    0.7744    0.1026    0.4268    0.2863   0.02076  0.005651   0.01536
      5/49     26.6G   0.02793  0.007966   0.01677   0.05267        99       640    0.7443    0.2085    0.5742    0.3594   0.01942   0.00517   0.01323
      6/49     26.6G   0.02725  0.007652   0.01521   0.05011       124       640    0.8961    0.4111    0.6982    0.4662   0.01891  0.004763    0.0112
      7/49     26.6G   0.02647  0.007082   0.01335    0.0469        60       640    0.8243    0.5997    0.7712    0.5037   0.01983  0.004469  0.009233
      8/49     26.6G   0.02589  0.006675   0.01197   0.04454       100       640    0.7525    0.7447    0.8206    0.5269   0.01894  0.004239  0.007568
      9/49     26.6G   0.02532  0.006556   0.01036   0.04224        97       640    0.6821    0.8476     0.846    0.5418   0.01888  0.004064  0.006214
     10/49     26.6G   0.02439  0.006173  0.009521   0.04009       127       640    0.6374    0.8982    0.8584    0.5549   0.01873   0.00394    0.0052
     11/49     26.6G   0.02383  0.005892  0.008473    0.0382       102       640    0.6565    0.9025    0.8707    0.5842   0.01689  0.003834  0.004585
     12/49     26.6G   0.02368  0.005977   0.00798   0.03764        76       640    0.6294    0.9136    0.8725    0.5743   0.01711  0.003778  0.003994
     13/49     26.6G   0.02296  0.005889  0.007465   0.03632        86       640    0.6256    0.9192    0.8742    0.5822   0.01707  0.003736  0.003595
     14/49     26.6G   0.02303  0.005677    0.0074   0.03611        95       640    0.6391    0.9184    0.8824    0.5992   0.01672  0.003699  0.003284
     15/49     26.6G   0.02237  0.005711  0.006903   0.03498        76       640    0.6301    0.9192    0.8829    0.6178   0.01643  0.003675  0.003101
     16/49     26.6G   0.02244  0.005626  0.006771   0.03483       122       640    0.6185    0.9185    0.8795    0.6112   0.01721  0.003638  0.002892
     17/49     26.6G   0.02172  0.005632  0.006372   0.03372       110       640    0.6175    0.9233    0.8819    0.6065   0.01665  0.003666  0.002729
     18/49     26.6G   0.02172  0.005484  0.006247   0.03345        90       640    0.6321    0.9186    0.8849    0.6228   0.01624   0.00362  0.002583
     19/49     26.6G   0.02146   0.00545  0.006128   0.03303        84       640    0.6467    0.9196    0.8852    0.6328   0.01586  0.003584   0.00253
     20/49     26.6G   0.02141  0.005508   0.00608     0.033       128       640    0.6466    0.9219    0.8882    0.6297     0.016  0.003567  0.002428
     21/49     26.6G   0.02092  0.005503  0.005755   0.03218        97       640    0.6448    0.9214    0.8875    0.6445   0.01563  0.003571  0.002367
     22/49     26.6G   0.02122  0.005489  0.005977   0.03268       101       640    0.6455    0.9254    0.8909    0.6486   0.01547  0.003554  0.002277
     23/49     26.6G   0.02091   0.00538  0.005707     0.032        67       640    0.6563     0.923    0.8916    0.6378   0.01566  0.003523  0.002176
     24/49     26.6G   0.02088  0.005567  0.005509   0.03195        94       640      0.65    0.9264    0.8918    0.6502   0.01549  0.003529  0.002116
     25/49     26.6G   0.02058  0.005351  0.005373   0.03131        87       640    0.6578    0.9265    0.8929    0.6566   0.01507  0.003516  0.002059
     26/49     26.6G   0.02036  0.005376  0.005379   0.03111        80       640    0.6508    0.9264    0.8913     0.655   0.01539  0.003513   0.00203
     27/49     26.6G   0.02037  0.005328  0.005293   0.03099        89       640    0.6591    0.9233    0.8924    0.6567   0.01512  0.003491  0.001966
     28/49     26.6G   0.01998  0.005309  0.005264   0.03055        72       640    0.6506    0.9238    0.8905    0.6582   0.01512  0.003506  0.001909
     29/49     26.6G   0.01997   0.00539  0.005175   0.03053        83       640    0.6463    0.9248     0.892    0.6594   0.01517  0.003503  0.001862
     30/49     26.6G   0.02009  0.005301   0.00524   0.03063        79       640    0.6604    0.9226    0.8929    0.6657   0.01494  0.003478  0.001848
     31/49     26.6G   0.01971  0.005228   0.00493   0.02987        94       640    0.6551    0.9226    0.8925    0.6663   0.01488  0.003481  0.001794
     32/49     26.6G   0.01966  0.005202  0.004975   0.02984        98       640    0.6574    0.9211    0.8925    0.6668   0.01487  0.003477  0.001784
     33/49     26.6G   0.01966  0.005309  0.005166   0.03014        84       640    0.6562    0.9222    0.8931    0.6682   0.01476  0.003477  0.001722
     34/49     26.6G   0.01942  0.005193  0.004892    0.0295       109       640    0.6636    0.9223    0.8944    0.6681   0.01486  0.003477   0.00171
     35/49     26.6G   0.01921  0.005199  0.004987    0.0294        99       640    0.6634    0.9223    0.8944    0.6694   0.01481  0.003477  0.001666
     36/49     26.6G   0.01923  0.005343  0.005028    0.0296        99       640      0.66    0.9244    0.8935    0.6697   0.01477  0.003479  0.001657
     37/49     26.6G   0.01898  0.005209  0.004676   0.02886       116       640    0.6624     0.922    0.8934    0.6727   0.01462   0.00346  0.001639
     38/49     26.6G   0.01909  0.005219  0.004868   0.02918       101       640     0.666    0.9218    0.8943    0.6724   0.01471  0.003464   0.00163
     39/49     26.6G   0.01906  0.005007  0.004686   0.02875        80       640    0.6622    0.9234    0.8938    0.6726   0.01468  0.003469  0.001598
     40/49     26.6G   0.01845  0.005085  0.004513   0.02805        83       640    0.6644    0.9215    0.8941    0.6736   0.01462  0.003462  0.001558
     41/49     26.6G   0.01851  0.005162  0.004566   0.02824        68       640    0.6643    0.9241     0.895    0.6732   0.01476  0.003472   0.00154
     42/49     26.6G    0.0188   0.00513  0.004638   0.02857       113       640    0.6616    0.9256    0.8952    0.6741   0.01458   0.00347  0.001517
     43/49     26.6G   0.01857  0.005158  0.004722   0.02845        82       640     0.665    0.9252    0.8951    0.6742   0.01469  0.003471  0.001513
     44/49     26.6G   0.01842   0.00503  0.004532   0.02798        81       640    0.6641    0.9248     0.896    0.6759   0.01462  0.003466  0.001489
     45/49     26.6G    0.0189   0.00518  0.004527    0.0286        78       640    0.6641    0.9265    0.8957    0.6758   0.01469  0.003466  0.001479
     46/49     26.6G   0.01864  0.005219  0.004669   0.02853        83       640    0.6576     0.927    0.8962    0.6765   0.01463  0.003466  0.001472
     47/49     26.6G   0.01828  0.005248  0.004482   0.02801        90       640      0.66    0.9262    0.8958    0.6758   0.01462  0.003466  0.001464
     48/49     26.6G   0.01856  0.005147  0.004514   0.02822       102       640    0.6613    0.9265    0.8959    0.6756   0.01457  0.003466  0.001459
     49/49     26.6G   0.01839  0.005028  0.004471   0.02789        79       640    0.6633    0.9244    0.8956    0.6766   0.01459  0.003458  0.001451

**Comment by glenn-jocher** [comment]
@abhiagwl4262 your training looks fine. If anything, you should train longer. Default is 300 epochs.

**Comment by abhiagwl4262** [comment]
@glenn-jocher 
1. For warmup you chose max(2 epoches, 1000 iteration). How did you chose number of max iterations?
2. You are increasing your learning rate for  max(2 epoches, 1000 iteration) and then decreasing using cosine annealing till 300 epoches? Did you experiment with other learning rate strategies too?

**Comment by glenn-jocher** [comment]
@abhiagwl4262 we chose the current warmup period for stability and output bias training using trial and error. The cosine LR strategy in place provides improved performance over the original step LR policy used in YOLOv3. the max() operation is in place to force a longer warmup period for very small datasets like coco128.

**Comment by PariyantoNoviansyah** [comment]
> @pfeatherstone a random crop is chosen without consideration of object locations. Different datasets may benefit to different degrees from mosaic, so you may want to try training with and without it on your custom dataset. We've recently added mosaic to the training hyps next to mixup, so now it's easy to control the probability of any given image being mosaiced:
> 
> https://github.com/ultralytics/yolov5/blob/5a9c5c1d3b02c8796e6af153d57af5821d8db0d6/data/hyp.scratch.yaml#L32-L33

**Comment by PariyantoNoviansyah** [comment]
> @pfeatherstone a random crop is chosen without consideration of object locations. Different datasets may benefit to different degrees from mosaic, so you may want to try training with and without it on your custom dataset. We've recently added mosaic to the training hyps next to mixup, so now it's easy to control the probability of any given image being mosaiced:
> 
> https://github.com/ultralytics/yolov5/blob/5a9c5c1d3b02c8796e6af153d57af5821d8db0d6/data/hyp.scratch.yaml#L32-L33

@glenn-jocher where did you get the mosaic value and what is meant by probability, is it a unit of value?

**Comment by glenn-jocher** [comment]
@PariyantoNoviansyah mosaic probability (0-1) is the chance that a mosaic will be used for a given image. If you use 0.5 for example then 50% of the images will use mosaics on average.
